odl/trafos/fourier.py

```diff
@@ -1645,6 +1647,211 @@ def inverse(self):
+            ):
+            raise ValueError('One of the samples in `non_uniform_samples` is'
+              ' not a float between -0.5 and 0.5')
+
+    def _is_sample_valid(self, sample):
+        """Check whether a single sample is correct.
+
+        Parameters
+        ----------
+        sample : iterable
+            The sample whose location you want to check
+
+        Returns
+        -------
+        bool : Whether the sample contains only floats in [-0.5; 0.5[
+        """
+        floats = all(isinstance(coord, float) for coord in sample)
+        if floats:
+            normalized = all(-0.5 <= coord < 0.5 for coord in sample)
+        else:
+            normalized = False
+        return floats & normalized
+
+    def _normalize(self, x):
+        """Normalize the result of the non uniform FFT.
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            The results of the non uniform FFT
+
+        Returns
+        -------
+        out : `numpy.ndarray`
+            Normalized results of the non uniform FFT
+        """
+        out = x / np.sqrt(self.nfft.M)
+        return out
+
+    @property
+    def adjoint(self):
+        """Adjoint of this instance."""
+        if self.adjoint_class:
+            return self.adjoint_class(
+                shape=self.shape,
+                non_uniform_samples=self.non_uniform_samples,
+            )
+        raise NotImplementedError(
+            "Adjoint not implemented for this non-uniform fourier operator",
+        )
+
+
+class NonUniformFourierTransform(NonUniformFourierTransformBase):
+    """Forward Non uniform Fast Fourier Transform.
+    """
+    def __init__(self, shape, non_uniform_samples):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : tuple
+            The dimensions of the data whose non uniform FFT you want to
+            compute
+        non_uniform_samples : iterable
+            List of the fourier space positions where the coefficients are
+            computed. Each position must be in [-0.5; 0.5[
+        """
+        super(NonUniformFourierTransform, self).__init__(
+            shape=shape,
+            non_uniform_samples=non_uniform_samples,
+            domain=discr_sequence_space(shape, dtype=np.complex128),
+            range=discr_sequence_space(
+                [len(non_uniform_samples)],
+                dtype=np.complex128,
+            ),
+        )
+        self.adjoint_class = NonUniformFourierTransformAdjoint
+
+    def _call(self, x):
+        """Compute the direct non uniform FFT.
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            The data whose non uniform FFT you want to compute
+
+        Returns
+        -------
+        out_normalized : `numpy.ndarray`
+            Result of the transform
+        """
+        self.nfft.f_hat = np.array(x)
+        out = self.nfft.trafo()
```

Comment:
Does NFFT support pre-allocated `out`? If yes, would be worth using.

Reply:
I haven't tried, but if so, why is it worth using?

Reply:
Because it saves a copy. `out = self.nfft.trafo()` allocates new memory for `out`, whereas (hypothetically) `self.nfft_trafo(out=out)` just updates `out`, which should be faster.
Caveat: the implementation may be crappy and the latter may actually be slower than the former. Ideally you would benchmark the two variants.

Silly example (run in IPython):
```python
>>> x = np.zeros(int(1e7))
>>> %timeit out = np.exp(x)
79.6 ms ± 174 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> out = np.empty_like(x)
>>> %timeit np.exp(x, out=out)
55.8 ms ± 309 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
I don't get the whole picture most certainly, but without any context, wouldn't a more fair benchmark be:

```
In [1]: import numpy as np                                                                                                                     

In [2]: x = np.zeros(int(1e7))                                                                                                                 

In [3]: %timeit out = np.exp(x)                                                                                                                
66.5 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit np.exp(x, out=np.empty_like(x))                                                                                                
67.3 ms ± 172 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Maybe it's because in this case there is no copy but not sure. But anyway, I don't think `pynfft` supports pre-allocated `out` so I don't think we should delve into this too much (neither from the function `trafo` nor by setting `nfft.f` in advance in the case of the direct transform).

Reply:
The two versions that you compare should be practically indistinguishable.

My point was that you can make use of `out` if the called function supports it, and if it's implemented well it will not copy and typically run faster.
Often that's interesting when an operation is part of a chain of operations. For instance, if you want to compute `log(x / y)` there are basically two possibilities (assuming it's okay to mutate `x`):

- `z = np.log(x / y)` -- more readable, less efficient
  Actually this is equivalent to
  ```python
  >>> tmp1 = np.divide(x, y)  # new array
  >>> z = np.log(tmp1)  # new array
  ```
- `x /= y; np.log(x, out=x)` -- less readable, more efficient
  For analogy, you can also write this as
  ```python
  >>> np.divide(x, y, out=x)  # no new array
  >>> np.log(x, out=x)  # no new array
  ```

You will see a lot of code like the latter spread throughout ODL.

NB: With NumPy [temporary elision](https://docs.scipy.org/doc/numpy/release.html#temporary-elision) things are no longer that clear-cut, but I don't know how much of a difference it makes.



> But anyway, I don't think `pynfft` supports pre-allocated `out`

Hm, that would be worth adding



> neither from the function `trafo` nor by setting `nfft.f` in advance in the case of the direct transform -this I just tested -

Thanks for checking!